### ISSUE TITLE
transmute_ptr_to_ptr: False positive when using transmute in consts

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust-clippy#5959]
Using `transmute` in `const`s (recently stabilized) to cast one pointer to another raises `transmute_ptr_to_ptr` warning, despite the fact that dereferencing raw pointers in constants is unstable.

```rust
const _: &() = {
    struct ZST;
    let zst = &ZST;

    unsafe {
        core::mem::transmute::<&'static ZST, &'static ()>(zst)   
    }
};
```
```
warning: transmute from a reference to a reference
 --> src/lib.rs:6:9
  |
6 |         core::mem::transmute::<&'static ZST, &'static ()>(zst)   
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(zst as *const _::ZST as *const ())`
  |
  = note: `#[warn(clippy::transmute_ptr_to_ptr)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#transmute_ptr_to_ptr
```
([playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=e8200c32de0e334f932dfaf7187df76c))
```rust
const _: &() = {
    struct ZST;
    let zst = &ZST;

    unsafe {
        &*(zst as *const ZST as *const ())
    }
};
```
```
error[E0658]: dereferencing raw pointers in constants is unstable
 --> src/lib.rs:6:9
  |
6 |         &*(zst as *const ZST as *const ())
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #51911 <https://github.com/rust-lang/rust/issues/51911> for more information
```
([playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=d1c529df3a6b7ac0810a95733599c8ce))

### Meta

- `cargo clippy -V`: clippy 0.0.212 (cfbc6d4 2020-08-21)
- `rustc -Vv`:
  ```
  rustc 1.46.0-beta.5 (cfbc6d4cf 2020-08-21)
  binary: rustc
  commit-hash: cfbc6d4cf1acbac5c025acf8ed68929caa551df3
  commit-date: 2020-08-21
  host: x86_64-unknown-linux-gnu
  release: 1.46.0-beta.5
  LLVM version: 10.0
  ```
